File: bongf/java/src/level2/FunctionDevelopment.java
package level2;

import java.util.Arrays;
import java.util.Stack;

public class FunctionDevelopment {
    public static void main(String[] args) {
        int[] progresses = {93, 30, 55};
        int[] speeds = {1, 30, 5};
        System.out.println(Arrays.toString(solution2(progresses, speeds)));
    }

    public static int[] solution(int[] progresses, int[] speeds) {
        Stack<Integer> answer = new Stack<>();
        Stack<Integer> stack = new Stack<>();
        stack.push(0);

        for (int i = 0; i < progresses.length; i++) {
            int p = progresses[i];
            int s = speeds[i];
            int d = (100 - p) % s == 0 ? (100 - p) / s : (int) ((100 - p) / s) + 1;

            if (stack.peek() < d) {
                stack.pop();
                stack.push(d);
                answer.push(1);
            } else {
                answer.push(answer.pop() + 1);
            }
        }
        return answer.stream().mapToInt(Integer::valueOf).toArray();
    }

    public static int[] solution2(int[] progresses, int[] speeds) {
        int[] dayOfEnd = new int[100];
        int day = -1;
        for(int i=0; i<progresses.length; i++) {
            while(progresses[i] + (day*speeds[i]) < 100) {
                day++;
            }
            dayOfEnd[day]++;
        }
        return Arrays.stream(dayOfEnd).filter(i -> i!=0).toArray();
    }
}

File: programmers/python/level2/function_development.py
import math 

def solution(progresses, speeds):
    days = []
    for p, s in zip(progresses, speeds) :
        days.append((100-p) // s + 1 if ((100 - p) % s) != 0 else (100-p) // s)
    
    answer = [0]
    criteria = days[0]
    for d in days :
        if criteria >= d:
            answer[-1] += 1
        else :
            criteria = d
            answer.append(1)       
            
    return answer

## 작업에 필요할 날과 그것을 정답 배열과 비교해 주는 것을 한 번에 
def solution2(progresses, speeds):
    Q=[]
    for p, s in zip(progresses, speeds):
        if len(Q)==0 or Q[-1][0]<-((p-100)//s):
            Q.append([-((p-100)//s),1])
        else:
            Q[-1][1]+=1
    return [q[1] for q in Q]

print(solution([93, 30, 55], [1, 30, 5]))
